src/server/shell/basic_accessibility_manager.h

```diff
@@ -68,23 +68,42 @@ class BasicAccessibilityManager : public AccessibilityManager
     void max_speed(double x_axis, double y_axis) override;
 
 private:
-    class MousekeyPointer;
-
     struct MutableState {
         // 25 rate and 600 delay are the default in Weston and Sway
         int repeat_rate{25};
         int repeat_delay{600};
 
         std::vector<std::shared_ptr<shell::KeyboardHelper>> keyboard_helpers;
-        std::shared_ptr<MousekeyPointer> mousekey_pointer;
+    };
+
+    template <typename Transformer> class Registration
+    {
+    public:
+        Registration(
+            std::shared_ptr<Transformer> const& transformer,
+            std::shared_ptr<input::InputEventTransformer> const& event_transformer);
+
+        void add_registration();
+        void remove_registration();
+
+        Transformer* operator->() const noexcept;
+
+        bool is_registered() const noexcept;
+
+    private:
+        Registration(Registration const&) = delete;
+        Registration& operator=(Registration const&) = delete;
+
+        std::shared_ptr<Transformer> const transformer;
+        std::shared_ptr<input::InputEventTransformer> const event_transformer;
+        bool registered{false};
```

Comment:
This mutable state is accessed without synchronization. This is probably a case for `std::atomic<bool>` and `compare_exchange_strong()`